src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -0,0 +1,142 @@
+        // Recursively apply transformation for sub-graph based operations
+        if (auto sub_graph_node = std::dynamic_pointer_cast<ov::op::util::SubGraphOp>(op)) {
+            if (auto sub_graph = sub_graph_node->get_function()) {
+                rewritten = shared_node_optimization(sub_graph, rules) || rewritten;
+            }
+        }
+        for (auto& output : op->outputs()) {
+            const auto& target_inputs = output.get_target_inputs();
+            if (target_inputs.size() <= 1)
+                continue;  // nothing to optimize
+            std::unordered_map<ov::Node::type_info_t, std::vector<ov::Node*>> type_to_node;
+            for (const auto& input : target_inputs) {
+                auto node = input.get_node();
+                if (node && rules.count(node->get_type_info()))
+                    type_to_node[node->get_type_info()].push_back(node);
+            }
+            for (auto& item : type_to_node) {
+                const auto& shared_nodes = item.second;
+                if (shared_nodes.size() < 2)
+                    continue;
+                const auto& ops_type = item.first;
+                const auto& are_equal = rules.at(ops_type);
+                const auto& root_op = shared_nodes[0];
+                for (const auto& child_op : shared_nodes) {
+                    if (root_op->get_instance_id() != child_op->get_instance_id() && are_equal(root_op, child_op)) {
+                        rewritten = replace_output_update_name(child_op->output(0), root_op->output(0)) || rewritten;
+                    }
+                }
+            }
+        }
+    }
+    return rewritten;
+}
+
+bool inputs_from_same_source_or_equal_constants(const Node* lhs, const Node* rhs) {
+    if (lhs->get_input_size() != rhs->get_input_size())
+        return false;
+    size_t input_size = lhs->get_input_size();
+    for (size_t i = 0; i < input_size; ++i) {
+        if (lhs->input_value(i) == rhs->input_value(i))
+            continue;
+        OPENVINO_SUPPRESS_DEPRECATED_START
+        auto lhs_constant = ov::get_constant_from_source(lhs->input_value(i));
+        auto rhs_constant = ov::get_constant_from_source(rhs->input_value(i));
+        OPENVINO_SUPPRESS_DEPRECATED_END
+        if (!lhs_constant || !rhs_constant)
+            return false;
+        if (lhs_constant->get_shape() != rhs_constant->get_shape() ||
+            ov::shape_size(lhs_constant->get_shape()) > 10)  // protection for not comparing huge constants
+            return false;
+        if (lhs_constant->get_element_type() != rhs_constant->get_element_type())
+            return false;
+        if (lhs_constant->get_element_type().is_integral()) {
+            if (lhs_constant->cast_vector<int64_t>() != rhs_constant->cast_vector<int64_t>())
+                return false;
+        } else if (lhs_constant->get_element_type().is_real()) {
+            if (lhs_constant->cast_vector<double>() != rhs_constant->cast_vector<double>())
```

Comment:
Since type equality is asserted these `cast`s bring no value. Look at https://github.com/openvinotoolkit/openvino/pull/17694/files#diff-e6fd768fedc8b143cfb95dfe0eefd599ed0cc5e3455767955d64707819bc7ef4